### ISSUE TITLE
Mock and verify call when using dart extensions

### ISSUE THREAD

**Issue by SandroMaglione** [dart-lang/mockito#239]
I am trying to verify a call of a method inside my class. The problem is that the method called is part of a custom dart extension I defined. Therefore, it is not recognized by Mockito as a valid mocked method.

For example, my situation is:
```dart
test('should check internet connection before making any remote request',
        () async {
      when(connectionCheckerMock.hasInternetConnection(any))
          .thenAnswer((_) async => true);
      await loginRepositoryMock.signInEmail(tEmail, tPassword);
      verify(connectionCheckerMock.hasInternetConnection(any)).called(1);
    });
```
The problem is that `hasInternetConnection` is defined inside an extension and it has a non-mockito implementation.

The package throws these errors:
```
Invalid argument(s): An argument matcher (like `any`) was used as a named argument, but did not use a Mockito "named" API. Each argument matcher that is used as a named argument needs to specify the name of the argument it is being used in. For example: `when(obj.fn(x: anyNamed("x")))`.
```
```
Bad state: Mock method was not called within `when()`. Was a real method called?
```

Am I doing something wrong? Is there a way to verify the method call in this situation when using extensions?

**Comment by srawlins** [comment]
It sounds like we would benefit from a few things:

* Document mockito + extensions
* Provide a better error in the case you show; I don't think this is possible today :(

You can solve your issue, very awkwardly by doing the following:

```dart
class ConnectionChecker {}

extension X on ConnectionChecker {
  bool hasInternetConnection(Object arg1) => ...;
}

class MockConnectionChecker extends Mock implements ConnectionChecker {
  // Hand-written stub. Super unfortunate.
  bool hasInternetConnection(Object arg1) => noSuchMethod(
      Invocation.method(#hasInternetConnection, [arg1]));
}
```

This should work because instance methods take precedence over extension methods. That stub takes the extension method out of the equation. This would not work if the extension method is [invoked explicitly](https://github.com/dart-lang/language/blob/master/accepted/2.6/static-extension-members/feature-specification.md#explicit-extension-member-invocation), via `X(connectionCheckerMock).hasInternetConnection(any)`.

**Comment by SandroMaglione** [comment]
For any reference/help, this is the full code example:

**ConnectionChecker**
```dart
abstract class ConnectionChecker {
  /// Check if a valid internet connection is available
  Future<bool> get isConnected;
}
```

**Extension**
```dart
extension ErrorWhenNoInternetConnection on ConnectionChecker {
  void hasInternetConnection(String message) async {
    if (!(await this.isConnected)) throw ConnectionError(message);
  }
}
```

**Called method** (inside `LoginRepositoryImpl`)
```dart
Future<User> signInEmail(String email, String password) async {
        await connectionChecker.hasInternetConnection(
            'Impossbile to sign in, no internet connection available');
        ...
```
**Mock test**
```dart
setUp(() {
  connectionCheckerMock = ConnectionCheckerMock();
  loginRepository = LoginRepositoryImpl(
    connectionChecker: connectionCheckerMock,
  );
});

group('signInEmail', () {
  test('should check internet connection before making any remote request',
      () async {
    when(connectionCheckerMock.hasInternetConnection(any))
        .thenAnswer((_) async => true);
    await loginRepository.signInEmail(tEmail, tPassword);

    /// Verify call on extension method!
    verify(connectionCheckerMock.hasInternetConnection(any)).called(1);
  });
});
```

**Comment by natebosch** [comment]
**tl;dr;** - you cannot mock extension methods at all - if you need to be able to override it for a test move it to a separate class as an instance method.

**details**
Only instance methods may be mocked - not static methods of any flavor. Class statics, top level methods, extension methods are all static.

> This should work because instance methods take precedence over extension methods.

This is not quite true. _statically known_ instance methods take precedence over extension methods. Calling an extension method is like calling a top level method - it goes straight to the method and doesn't care about the instance. Adding a runtime check to see if an instance happens to have a member we didn't know about statically and picking at runtime between the instance or extension would be too expensive.

The choice is made at the call site. You can know exactly which you'll get by using "jump to definition" in your IDE. If you get to a class member, you'll get an instance method. If you get to an extension it will _aways_ call that extension.

Here is an example:

```dart
void main() {
  print(takeFoo(Foo())); // 'extension'
  print(takeFoo(ExtraFoo())); // 'extension'
  print(takeFoo(ExtraExtraFoo())); // 'extension'
​
  print(takeExtraFoo(ExtraFoo())); // 'a'
  print(takeExtraFoo(ExtraExtraFoo())); // 'b'
}
​
String takeFoo(Foo foo) {
  return foo.extra; // Always the extension getter. Statically resolved
}
​
String takeExtraFoo(ExtraFoo foo) {
  return foo.extra; // Always an instance method.
}
​
class Foo {}
​
extension on Foo {
  String get extra => 'extension';
}
​
class ExtraFoo extends Foo {
  String get extra => 'a';
}
​
class ExtraExtraFoo extends ExtraFoo {
  String get extra => 'b';
}
```

Closing since there isn't anything we can do in mockito about this. Anything that needs to be mocked in your test needs to be designed to be mockable - that means only using instance methods for things that need to be mocked.
